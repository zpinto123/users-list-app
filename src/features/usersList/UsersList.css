
.container{
  width: 450px;
  justify-content: center;
  align-items: center;
  text-align: center;
}

.usersList {
  border-radius: 5px;
}

.searchContainer {
  margin-bottom: 10px;
}