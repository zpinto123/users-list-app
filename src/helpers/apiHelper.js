import api from "../services/config";

const STATUS_OK = 200;

const fetchData = (url, method = "GET") =>
  fetch(url, { method, headers: setAuthHeaders() })
    .then(verifyResponse)
    .catch(handleErrors);

const verifyResponse = res => {
  if (!res || res.status !== STATUS_OK) {
    console.log(`error fetching - status: ${res.status}`);
    return res.status;
  }
  return res.json();
};

const handleErrors = error => console.error(error);

const setAuthHeaders = () => {
  const { username, password } = api;
  const headers = new Headers();
  headers.set(
    "Authorization",
    `Basic ${Buffer.from(`${username}:${password}`).toString("base64")}`
  );
  return headers;
};

export { fetchData };
