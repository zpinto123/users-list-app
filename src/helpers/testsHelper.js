import configureStore from "redux-mock-store";
import initialState from "../redux/modules/users/initialState";

const mockStore = configureStore();
const store = mockStore(initialState);

export { store };
