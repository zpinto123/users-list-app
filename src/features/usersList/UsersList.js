import React, { PureComponent } from "react";
import PropTypes from "prop-types";

import messages from "./messages";

import { Title, TextBox } from "../../common";
import UsersListItem from "./parts/usersListItem";

import "./UsersList.css";

class UsersList extends PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      selectedUserId: null,
      searchUser: ""
    };
  }

  componentWillMount = () => this.props.getUsersList();

  renderErrorMessage = () => (
    <div>
      <span>{messages.errors.userListFetchFail()}</span>
    </div>
  );

  handleClick = id => () => {
    const { selectedUserId } = this.state;
    this.setState({ selectedUserId: selectedUserId === id ? null : id });
  };

  handleUserSearch = ({ target: { value } }) =>
    this.setState({ searchUser: value.toLowerCase(), selectedUserId: null });

  render() {
    const { usersList, error } = this.props;
    const { selectedUserId, searchUser } = this.state;

    return (
      <div className="container">
        <Title text="Users list" />
        {error ? (
          this.renderErrorMessage()
        ) : (
          <div className="usersList">
            <div>
              <div className="searchContainer">
                <TextBox
                  placeholder="Search by surname"
                  handleChange={this.handleUserSearch}
                />
              </div>
              {usersList &&
                usersList
                  .filter(({ surname }) =>
                    surname.toLowerCase().includes(searchUser)
                  )
                  .map(user => (
                    <UsersListItem
                      key={user.id}
                      userInfo={user}
                      isSelected={selectedUserId === user.id}
                      handleClick={this.handleClick(user.id)}
                    />
                  ))}
            </div>
          </div>
        )}
      </div>
    );
  }
}

UsersList.propTypes = {
  usersList: PropTypes.array.isRequired,
  error: PropTypes.bool.isRequired
};

export default UsersList;
