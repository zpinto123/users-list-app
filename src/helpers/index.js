import { fetchData } from "./apiHelper";

export { fetchData };
