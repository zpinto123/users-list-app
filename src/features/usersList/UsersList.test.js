import React from "react";
import ShallowRenderer from "react-test-renderer/shallow";

import { store } from "../../helpers/testsHelper";
import UsersList from "./UsersList";

describe("<UsersList />", () => {
  const list = [
    {
      id: "8sKid80",
      name: "JosÃ© Pinto",
      surname: "Pinto",
      email: "super@super.com",
      employer: "Google"
    },
    {
      id: "8sKids80",
      name: "John Sila",
      surname: "Sila",
      email: "supera@supera.com",
      employer: "Microsoft"
    }
  ];
  const error = false;

  it("renders UsersList correctly", () => {
    const mockGetUsersList = jest.fn();
    const renderer = new ShallowRenderer();
    const component = renderer.render(
      <UsersList
        store={store}
        getUsersList={mockGetUsersList}
        usersList={list}
        error={error}
      />
    );
    expect(component).toMatchSnapshot();
  });
});
