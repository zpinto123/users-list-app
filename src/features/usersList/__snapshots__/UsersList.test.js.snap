// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UsersList /> renders UsersList correctly 1`] = `
<div
  className="container"
>
  <Title
    text="Users list"
  />
  <div
    className="usersList"
  >
    <div>
      <div
        className="searchContainer"
      >
        <TextBox
          handleChange={[Function]}
          placeholder="Search by surname"
          type="text"
        />
      </div>
      <UsersListItem
        handleClick={[Function]}
        isSelected={false}
        userInfo={
          Object {
            "email": "super@super.com",
            "employer": "Google",
            "id": "8sKid80",
            "name": "JosÃ© Pinto",
            "surname": "Pinto",
          }
        }
      />
      <UsersListItem
        handleClick={[Function]}
        isSelected={false}
        userInfo={
          Object {
            "email": "supera@supera.com",
            "employer": "Microsoft",
            "id": "8sKids80",
            "name": "John Sila",
            "surname": "Sila",
          }
        }
      />
    </div>
  </div>
</div>
`;
